File: app/src/main/java/com/mac/kt_base/s2/KtBase33.java
package com.mac.kt_base.s2;

interface IShowResult { // 接口的折中方案 解决 kt的lambda问题
    void result(String result);
}

// TODO 33.Kotlin语言的匿名函数与具名函数
public class KtBase33 {

    public static void main(String[] args) {
        // KtBase33.kt匿名函数相当于
        showPersonInfo("Derry", 38, '男', "学习KT语言中", new IShowResult() {
            @Override
            public void result(String result) {
                System.out.println("匿名函数 显示结果:" + result);
            }
        });

        System.out.println();

        // KtBase33.kt具名函数相当于
        IShowResult showResultImpl = new MShowResultImpl();
        showPersonInfo("Derry", 38, '男', "学习KT语言中", showResultImpl);
    }

    // 具名函数的区域
    static class MShowResultImpl implements IShowResult {
        @Override
        public void result(String result) {
            System.out.println("具名函数 显示结果:" + result);
        }
    }

    // ---------------------------------

    static void showPersonInfo(String name, int age, char sex, String study , IShowResult iShowResult) {
        String str = String.format("name:%s, age:%d, sex:%c, study:%s", name, age, sex, study);
        iShowResult.result(str);
    }
}

File: app/src/main/java/com/mac/kt_base/s2/KtBase33.kt
package com.mac.kt_base.s2

// TODO 33.Kotlin语言的匿名函数与具名函数
fun main() {
    // 匿名函数
    showPersonInfo("Derry", 38, '男', "学习KT语言中") {
        println("匿名函数 显示结果:$it")
    }

    println()

    // 具名函数
    showPersonInfo("Derry", 38, '男', "学习KT语言中", ::showResultImpl)
}

// showResultImpl 是 实实在在的函数名 所以叫具名
fun showResultImpl(it: String) {
    println("具名函数 显示结果:$it")
}


inline fun showPersonInfo(name: String, age: Int, sex: Char, study: String, showResult: (String) -> Unit) {
    val str = "name:$name, age:$age, sex:$sex, study:$study"
    showResult(str)
}